accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,871 @@
+        determined by downwards inference.
+
+### Type promotion
+
+An `ifElement` interacts with type promotion in the same way that `if`
+statements do. Given an `ifElement` with condition `condition`, then element
+`thenElement` and optional else element `elseElement`:
+
+*   If `condition` shows that a local variable *v* has type `T`, then the type
+    of *v* is known to be `T` in `thenElement`, unless any of the following are
+    true:
+
+    *   *v* is potentially mutated in `thenElement`,
+    *   *v* is potentially mutated within a function other than the one where
+        *v* is declared, or
+    *   *v* is accessed by a function defined in `thenElement` and
+    *   *v* is potentially mutated anywhere in the scope of *v*.
+
+*Note: The type promotion rules for `if` will likely get more sophisticated in a
+future version of Dart because of non-nullable types. When that happens, `if`
+elements should continue to match `if` statements.*
+
+### Compile-time errors
+
+After type inference and disambiguation, the collection is checked for other
+compile-time errors. It is a compile-time error if:
+
+*   The collection is a map literal and *leaf elements* contains any
+    `expressionElement` elements.
+
+    ```dart
+    <String, int>{"not entry"} // Error.
+    ```
+
+*   The collection is a set literal and *leaf elements* contains any `mapEntry`
+    elements.
+
+    ```dart
+    ["not": "expression"] // Error.
+    <String>{"not": "expression"} // Error.
+    ```
+
+    *We prohibit map entries in list literals syntacitcally, earlier in the
+    proposal.*
+
+*   The collection is a list and the type of any of the *leaf elements* may not
+    be assigned to the list's element type.
+
+    ```dart
+    <int>[if (true) "not int"] // Error.
+    ```
+
+*   The collection is a map and the key type of any of the *leaf elements* may
+    not be assigned to the map's key type.
+
+    ```dart
+    <int, int>{if (true) "not int": 1} // Error.
+    ```
+
+*   The collection is a map and the value type of any of the *leaf elements* may
+    not be assigned to the map's value type.
+
+    ```dart
+    <int, int>{if (true) 1: "not int"} // Error.
+    ```
+
+*   A spread element in a list or set literal has a static type that is `Null`,
+    or that is not `dynamic` and not a subtype of `Iterable<Object>`.
+
+*   A spread element in a list or set has a static type that implements
+    `Iterable<T>` for some `T` and `T` is not assignable to the element type of
+    the list.
+
+*   A spread element in a map literal has a static type that is `Null`, or that
+    is not `dynamic` and not a subtype of `Map<Object, Object>`.
+
+*   If a map spread element's static type implements `Map<K, V>` for some `K`
+    and `V` and `K` is not assignable to the key type of the map or `V` is not
+    assignable to the value type of the map.
+
+*   The variable in a `for` element (either `for-in` or C-style) is declared
+    outside of the element to be `final`.
```

Comment:
Or to not have a setter. It doesn't have to be a local variable :)

get x => 42;
main() {
 { for(x in [1]) x };
}
```
is an error, but `x` is not declared to be "final" anywhere.